cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go

```diff
@@ -290,6 +282,84 @@ func (s *Server) GetAvailablePackageDetail(ctx context.Context, request *corev1.
 	}, nil
 }
 
+// GetAvailablePackageVersions returns the package versions managed by the 'helm' plugin
+func (s *Server) GetAvailablePackageVersions(ctx context.Context, request *corev1.GetAvailablePackageVersionsRequest) (*corev1.GetAvailablePackageVersionsResponse, error) {
+
+	if request.GetAvailablePackageRef().GetContext().GetNamespace() == "" || request.GetAvailablePackageRef().GetIdentifier() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "No request AvailablePackageRef.Context provided")
+	}
+	contextMsg := fmt.Sprintf("(cluster=[%s], namespace=[%s])", request.AvailablePackageRef.Context.Cluster, request.AvailablePackageRef.Context.Namespace)
+	log.Infof("+helm GetAvailablePackageVersions %s", contextMsg)
+
+	namespace := request.AvailablePackageRef.Context.Namespace
+
+	// After requesting a specific namespace, we have to ensure the user can actually access to it
+	if err := s.hasAccessToNamespace(ctx, namespace); err != nil {
+		return nil, err
+	}
+
+	unescapedChartID, err := getUnescapedChartID(request.AvailablePackageRef.Identifier)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("Requesting chart '%s' (latest version) in ns '%s'", unescapedChartID, namespace)
+	chart, err := s.manager.GetChart(namespace, unescapedChartID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Unable to retrieve chart: %v", err)
+	}
+	return &corev1.GetAvailablePackageVersionsResponse{
+		PackageAppVersions: packageAppVersionsSummary(chart.ChartVersions),
+	}, nil
+}
+
+// packageAppVersionsSummary converts the model chart versions into the required version summary.
+func packageAppVersionsSummary(versions []models.ChartVersion) []*corev1.GetAvailablePackageVersionsResponse_PackageAppVersion {
+	pav := []*corev1.GetAvailablePackageVersionsResponse_PackageAppVersion{}
+
+	// Use a version map to be able to count how many major, minor and patch versions
+	// we have included.
+	version_map := map[int64]map[int64][]int64{}
```

Comment:
> Currently, it's a great approach and I don't foresee any better alternative. However, if we plan to continue using the existing approach (pre-fetching repos to a cache), perhaps we should delegate this logic to the DB (leveraging indexes), otherwise, this operation could be kinda time-consuming in O(repos_charts_versions). See what you think

Yeah, I'd be happy to update this to use direct DB queries once we've settled on the cache (it's possible with postgresql semver fields, not sure about redis etc.), but right now I'm just re-using the existing assetsvc code (ie. pulling the chart which includes all chart versions in order). Also, even without doing the processing in the DB, it doesn't need to be O(repos_charts_versions) given that we could terminate the loop once we had the max number of versions (haven't done, just pointing out that'd be trivial).